TPC: Update laser track drift velocity calibration

wiechula opened the pull request:
* add device to pre-filter laser track candiates
  (tpc-laser-track-filter)
* split device for
  - simple local processing (tpc-calib-laser-tracks)
  - distributed processing (tpc-laser-tracks-calibrator)
* new struct for calibration data storage
* drift velocity calibration separately for A-/C-Side
* code cleanup

wiechula committed:
TPC: Update laser track drift velocity calibration

* add device to pre-filter laser track candiates
  (tpc-laser-track-filter)
* split device for
  - simple local processing (tpc-calib-laser-tracks)
  - distributed processing (tpc-laser-tracks-calibrator)
* new struct for calibration data storage
* drift velocity calibration separately for A-/C-Side
* code cleanup

shahor02 commented on Detectors/TPC/calibration/include/TPCCalibration/CalibLaserTracks.h:
```diff
@@ -102,26 +169,41 @@ class CalibLaserTracks
-  std::vector<TimePair> mDVperTF;                                ///< drift velocity and time offset per TF
-  bool mWriteDebugTree{false};                                   ///< create debug output tree
+  int mTriggerPos{0};                                          ///< trigger position, if < 0 it treats it as the CE position
+  float mBz{0.5};                                              ///< Bz field in kGaus
```
to be sure: our nominal field in kGaus is 5 (0.5 in Tesla)

wiechula replied:
Absolutely, thanks! fixed.

wiechula committed:
TPC: several unrelated smaller fixes and improvements

wiechula committed:
TPC: fix treatment of sync offset in LinkZS data

Ideally, the sync offset is aligned to the largest offset within one TF.
This value can be found by looping over all data once but requires
access to the full TF.
Alternatively, an assumption about the largest sync offset can be made.

Some unrealted fixes are includes as well

shahor02 reviewed (commented):
@wiechula  looks fine, just  1 comment below

anerokhi commented on Detectors/TPC/calibration/include/TPCCalibration/CalibLaserTracks.h:
```diff
@@ -48,25 +82,39 @@ class CalibLaserTracks
+                                                    mZmatchPairsTFA{other.mZmatchPairsTFA},
+                                                    mZmatchPairsTFC{other.mZmatchPairsTFC},
+                                                    mZmatchPairsA{other.mZmatchPairsA},
+                                                    mZmatchPairsC{other.mZmatchPairsC},
```
Copy constructor resets `mFinalized`?

wiechula replied:
Thanks, fixed.

anerokhi replied:
There are problems with `CalibLaserTracks(const CalibLaserTracks&) = default;` because of a CUDA compiler or something like this?

wiechula replied:
The problem I get with default copy constructor is:
```
In file included from /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/memory:76:          
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/unique_ptr.h:962:34: error: call to implicitly-deleted copy constructor of 'o2::tpc::CalibLaserTracks'
    { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }                                                                                                                                                                                                                                                      
                                 ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                                                                                              
alicesw/sw/SOURCES/O2/O2-work/0/Detectors/Calibration/include/DetectorsCalibration/TimeSlot.h:40:25: note: in instantiation of function template specialization 'std::make_unique<o2::tpc::CalibLaserTracks, const o2::tpc::CalibLaserTracks &>' requested here                                       
      mContainer = std::make_unique<Container>(*src.getContainer());                                                                                                                                                                                                                                                          
                        ^                                                                                                                                                                                                                                                                                                     /usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/stl_algobase.h:719:18: note: in instantiation of member function 'o2::calibration::TimeSlot<o2::tpc::CalibLaserTracks>::operator=' requested here                                                   
            *--__result = std::move(*--__last);                                                                                                                                                                                                                                                                               
                        ^                                                                                                                                                                                                                                                                                                     
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/stl_algobase.h:760:24: note: in instantiation of function template specialization 'std::__copy_move_backward<true, false, std::random_access_iterator_tag>::__copy_move_b<o2::calibration::TimeSlot<o2::tpc::CalibLaserTracks> *, o2::calibration:
:TimeSlot<o2::tpc::CalibLaserTracks> *>' requested here                                                                                                        
                                       _Category>::__copy_move_b(__first,                                                                                                                                                                                                                                                     
...
...
...
```

Defining the copy constructor explicitly solved the problem.

anerokhi replied:
Ok. You have `std::unique_ptr<o2::utils::TreeStreamRedirector> mDebugStream;`. Which you ignore in the copy-constructor.
And which makes defaulted copy-constructor deleted.

wiechula replied:
Ahh right. Forgot about that. Yes, I wanted the keep the `TreeStreamRedirector` as `unique_ptr`, since it is only needed in debug mode.

wiechula commented:
@shahor02 , can this be merged? the macos CI output is empty...
